### PR TITLE
New commit log

### PR DESCRIPTION
**Pull request by mambisi**

### CODE REVIEW

**Comment by bkontur** [comment]
@mambisi 
I would like to move all the commit log stuff to directory, please:
`storage/src/commit_log/*`

this `storage/src/persistent/* ` was dedicated more to common/api/traits for storage

**Comment by mambisi** [comment]
Ohk

File: storage/src/persistent/commit_log/mod.rs [hunk 1]
```diff
@@ -16,23 +15,185 @@ use crate::persistent::schema::{CommitLogDescriptor, CommitLogSchema};
...
+        }
+        return buf;
+    }
+}
+
+pub struct CommitLog {
+    path: PathBuf,
+}
+
+impl CommitLog {
```
Comment (bkontur) [inline storage/src/persistent/commit_log/mod.rs:141]: @mambisi 
could you, please, provide also delete function, if possible?

File: storage/src/persistent/commit_log/mod.rs [hunk 2]
```diff
@@ -16,23 +15,185 @@ use crate::persistent::schema::{CommitLogDescriptor, CommitLogSchema};
...
+    #[inline]
+    pub fn append_msg<B: AsRef<[u8]>>(&mut self, payload: B) -> Result<u64, TezedgeCommitLogError> {
+        let mut writer = Writer::new(&self.path)?;
+        let offset = writer.write(payload.as_ref())?;
+        Ok(offset)
+    }
+
+    #[inline]
+    pub fn read(&self, from: usize, limit: usize) -> Result<MessageSet, TezedgeCommitLogError> {
+        let mut reader = Reader::new(&self.path)?;
```
Comment (bkontur) [inline storage/src/persistent/commit_log/mod.rs:156]: @mambisi 
do we need to create new instances `Reader::new/Writer:new`, all the time? it does lots of things in `new`

Comment (mambisi) [reply storage/src/persistent/commit_log/mod.rs:156]: No it doesn't but the BufWriter need to be drop for changes to take effect, we can make on reader, i will try it out

Comment (tizoc) [reply storage/src/persistent/commit_log/mod.rs:156]: @mambisi isn't calling `flush()` on the writer enough? The drop just falls `flush_buf` which is also called by `flush` https://doc.rust-lang.org/src/std/io/buffered/bufwriter.rs.html#414-421

Comment (mambisi) [reply storage/src/persistent/commit_log/mod.rs:156]: I think it will work

File: storage/src/persistent/commit_log/mod.rs [hunk 3]
```diff
@@ -16,23 +15,185 @@ use crate::persistent::schema::{CommitLogDescriptor, CommitLogSchema};
...
+
+impl Iterator for MessageSet {
+    type Item = Message;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let index = self.indexes.get(self.cursor)?;
+        let data = &self.buf[self.acc..(self.acc + index.uncompressed_data_length as usize)];
+        self.acc += index.uncompressed_data_length as usize;
+        self.cursor += 1;
+        Some(data.to_vec())
```
Comment (tizoc) [inline storage/src/persistent/commit_log/mod.rs:54]: I don't know how performance sensitive is this code, so maybe this doesn't matter: we are allocating a new vector each time here, is that necessary?

File: storage/src/persistent/commit_log/mod.rs [hunk 4]
```diff
@@ -16,23 +15,185 @@ use crate::persistent::schema::{CommitLogDescriptor, CommitLogSchema};
...
+    fn new(position: u64, compressed_data_length: u64, uncompressed_data_length: u64) -> Self {
+        Self {
+            position,
+            compressed_data_length,
+            uncompressed_data_length,
+        }
+    }
+    fn to_vec(&self) -> Vec<u8> {
+        let mut buf = vec![];
+        for b in &self.position.to_be_bytes() {
```
Comment (tizoc) [inline storage/src/persistent/commit_log/mod.rs:124]: Shorter and probably faster:

```rust
 buf.extend_from_slice(&self.position.to_be_bytes());
```

Comment (mambisi) [reply storage/src/persistent/commit_log/mod.rs:124]: Yeah, completely forgot about that function

Comment (mambisi) [reply storage/src/persistent/commit_log/mod.rs:124]: Thanks

File: storage/src/persistent/commit_log/mod.rs (hunk 3 above)
Comment (tizoc) [reply storage/src/persistent/commit_log/mod.rs:54]: Found https://docs.rs/streaming-iterator/ which would solve the issue.

But as I said, if performance is not an issue here, then it doesn't matter.

**Pull request closed without merging**